packages/opentelemetry-plugin-redis/src/redis.ts

```diff
@@ -0,0 +1,192 @@
+import { EventEmitter } from 'events';
+
+export class RedisPlugin extends BasePlugin<typeof redisTypes> {
+  static readonly COMPONENT = 'redis';
+  readonly supportedVersions = ['^2.6.0']; // should be >= 2.6.0
+
+  constructor(readonly moduleName: string) {
+    super();
+  }
+
+  protected patch() {
+    if (this._moduleExports.RedisClient) {
+      this._logger.debug(
+        'Patching redis.RedisClient.prototype.internal_send_command'
+      );
+      shimmer.wrap(
+        this._moduleExports.RedisClient.prototype,
+        'internal_send_command',
+        this._getPatchInternalSendCommand()
+      );
+
+      this._logger.debug('patching redis.create_stream');
+      shimmer.wrap(
+        this._moduleExports.RedisClient.prototype,
+        'create_stream',
+        this._getPatchCreateStream()
+      );
+
+      this._logger.debug('patching redis.createClient');
+      shimmer.wrap(
+        this._moduleExports,
+        'createClient',
+        this._getPatchCreateClient()
+      );
+    }
+    return this._moduleExports;
+  }
+
+  protected unpatch(): void {
+    if (this._moduleExports) {
+      shimmer.unwrap(
+        this._moduleExports.RedisClient.prototype,
+        'internal_send_command'
+      );
+      shimmer.unwrap(
+        this._moduleExports.RedisClient.prototype,
+        'create_stream'
+      );
+      shimmer.unwrap(this._moduleExports, 'createClient');
+    }
+  }
+
+  /**
+   * Patch internal_send_command(...) to trace requests
+   */
+  private _getPatchInternalSendCommand() {
+    const plugin = this;
+    return function internal_send_command(original: Function) {
+      return function internal_send_command_trace(
+        this: redisTypes.RedisClient & RedisPluginClientTypes,
+        cmd?: RedisCommand
+      ) {
+        const parentSpan = plugin._tracer.getCurrentSpan();
+
+        // New versions of redis (2.4+) use a single options object
+        // instead of named arguments
+        if (arguments.length === 1 && typeof cmd === 'object') {
+          const span = plugin._tracer.startSpan(
+            `${RedisPlugin.COMPONENT}-${cmd.command}`,
+            {
+              kind: SpanKind.CLIENT,
+              parent: parentSpan || undefined,
+              attributes: {
+                [AttributeNames.COMPONENT]: RedisPlugin.COMPONENT,
+                [AttributeNames.DB_STATEMENT]: cmd.command,
+              },
+            }
+          );
+
+          // Set attributes for not explicitly typed RedisPluginClientTypes
+          if (this.options) {
+            span.setAttributes({
+              [AttributeNames.PEER_HOSTNAME]: this.options.host,
+              [AttributeNames.PEER_PORT]: this.options.port,
+            });
+          }
+          if (this.address) {
+            span.setAttribute(
+              AttributeNames.PEER_ADDRESS,
+              `redis://${this.address}`
+            );
+          }
+
+          let spanEnded = false;
+          const endSpan = (err?: Error | null) => {
```

Comment:
Could this be extracted as a helper function at the top level of the module?

Reply:
Agreed. I moved it out. I had to get rid of the `spanEnded` check, but if the span is being ended more than once, its probably a bug that we should log out anyways.